In [ ]:
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.stats.outliers_influence import variance_inflation_factor


<p align="center">
    <img src="tu.png" width="800">
</p>

<div align="center">

# Факултет по приложна математика и информатика

## Курсов проект по дисциплина Математическа статистика
##### /Проектно задание № 17 /

## Специалност: „Анализ на данни“

##### Ръководител: проф. д-р Красимира Проданова

</div>

#### Изготвил:

Фак.№: 391224017

Група: 73

Име: Антон Милев

### Съдържание
1. Условия	
2. Теоретична обосновкa
3. Решаване на задачите със Python

# Условия

### Задача 1.

Направен е експеримент върху нов модел пътна настилка за определяне на спирачния път на определен вид тежкотоварни автомобили при различни скорости и влажност

|Мерна единица|изм. 1|изм. 2|изм. 3|изм. 4|изм. 5|изм. 6|
|---|---:|---:|---:|---:|---:|---:|
| Скорост V (км/ч) | 35 | 50 | 65 | 80 | 95 | 110 |
| Спирачен път S (м) | 16 | 26 | 41 | 62 | 88 | 119 |
| Влажност W (%) | 20 | 30 | 40 | 50 | 60 | 70 |

- Моделирайте с уравнение на регресия от вида: 
$ S = \beta_0 + \beta_1 \cdot V + \beta_2 \cdot W $

- Коментирайте адекватността на този модел
- Прогнозирайте спирачния път и негов 95% доверителен интервал, когато автомобилът се движи със 70 км/ч, а влажността е 45%

### Задача 2.

Направено е изследване, определящо месеците, в които консерви от даден вид остават годни за консумация. Три вида консерви са тествани в 4 типа складове

| Вид консерва | Склад I | Склад II | Склад III | Склад IV |
|---|---:|---:|---:|---:|
| A | 33.1 | 39.0 | 33.8 | 33.0 |
| B | 27.2 | 36.9 | 29.7 | 28.5 |
| C | 27.4 | 29.2 | 26.7 | 30.9 |

Използвайте ниво на значимост 0.05, за да тествате хипотезата за липса на различие:
- за различните видове консерви
- за различните типове склад



# Теоретична обосновка

### Задача 1.

Множествената линейна регресия моделира връзката между зависима променлива и множество независими променливи. Параметрите на модела се оценяват чрез метода на най-малките квадрати, като се минимизира сумата от квадратите на остатъците. Адекватността се оценява чрез коефициента на детерминация $(R^2)$ и F-тест, които измерват обяснителната сила на модела. Локалната значимост на всеки регресор се проверява чрез t-тест на съответния коефициент. Моделът позволява генериране на точкови прогнози и доверителни интервали за очакваната стойност при зададени стойности на независимите променливи.

### Задача 2.
Двуфакторният дисперсионен анализ (ANOVA) без повторения изследва едновременното влияние на два качествени фактора върху количествена променлива. Общата вариация в данните се разлага на компоненти, дължащи се на първия фактор, втория фактор и случайна вариация (остатък). Чрез F-критерий се проверяват нулевите хипотези за липса на ефект от всеки от двата фактора. Отхвърлянето на нулевата хипотеза при избрано ниво на значимост $\alpha=0.05$ означава статистически значимо влияние на съответния фактор върху зависимата променлива. Анализът помага да се установи дали наблюдаваните разлики са систематични или случайни.


# Решаване на задачите със Python

## Задача 1.

$S = \beta_0 + \beta_1 \cdot V + \beta_2 \cdot W$

Имаме че:
$W = \frac{2}{3}V - \frac{10}{3}$

Което значи че влажността не добавя нова информация, затова моделът може да не я използва

In [16]:
data = {
    'V': [35, 50, 65, 80, 95, 110],
    'W': [20, 30, 40, 50, 60, 70],
    'S': [16, 26, 41, 62, 88, 119]
}
df = pd.DataFrame(data)

X = df[['V', 'W']]
X = sm.add_constant(X)
y = df['S']

# Проверка за мултиколинеарност (VIF)
vif_data = pd.DataFrame()
vif_data["feature"] = X.columns
vif_data["VIF"] = [variance_inflation_factor(X.values, i) for i in range(len(X.columns))]

# Избор на модел
if vif_data[vif_data["feature"] == "W"]["VIF"].values[0] > 10:
    print("Открита е зависимост (мултиколинеарност). Използваме само скорост (V).")
    X_final = X[['const', 'V']]
else:
    print("Данните са независими. Използваме пълния модел с V и W.")
    X_final = X

model = sm.OLS(y, X_final).fit()
print(model.summary())

prediction = model.predict([1, 70])
print(f"\nПрогноза за 70 км/ч: {prediction[0]:.3f} метра")

Открита е зависимост (мултиколинеарност). Използваме само скорост (V).
                            OLS Regression Results                            
Dep. Variable:                      S   R-squared:                       0.966
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     112.7
Date:                Fri, 22 May 2026   Prob (F-statistic):           0.000446
Time:                        09:12:47   Log-Likelihood:                -19.871
No. Observations:                   6   AIC:                             43.74
Df Residuals:                       4   BIC:                             43.33
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------

c:\Users\Anton\anaconda3\Lib\site-packages\statsmodels\regression\linear_model.py:1782: RuntimeWarning: divide by zero encountered in scalar divide
  return 1 - self.ssr/self.centered_tss
c:\Users\Anton\anaconda3\Lib\site-packages\statsmodels\stats\outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)
c:\Users\Anton\anaconda3\Lib\site-packages\statsmodels\stats\stattools.py:74: ValueWarning: omni_normtest is not valid with less than 8 observations; 6 samples were given.
  warn("omni_normtest is not valid with less than 8 observations; %i "


### Разсъждение и обобщение:
Поради перфектната линейна зависимост между скоростта $(V)$ и влажността $(W)$, данните проявяват мултиколинеарност, което прави едновременното им включване математически нестабилно. Поради това е приложен еднофакторен модел:
$ \hat{S}=-41.04+1.38\cdot V $
Изводи:
- Адекватност: Моделът е статистически значим $(p < 0.001)$ и обяснява $96.6%$ от вариацията на спирачния път $(R^2=0.966)$
- Интерпретация: С всяко увеличение на скоростта с 1 км/ч, спирачният път се увеличава средно с 1.38 м.
- Прогноза: При 70 км/ч очакваният спирачен път е 55.23 м.


## Задача 2.

In [12]:
data = {
    'Вид': ['A', 'A', 'A', 'A', 'B', 'B', 'B', 'B', 'C', 'C', 'C', 'C'],
    'Склад': ['I', 'II', 'III', 'IV', 'I', 'II', 'III', 'IV', 'I', 'II', 'III', 'IV'],
    'Годност': [33.1, 39.0, 33.8, 33.0, 27.2, 36.9, 29.7, 28.5, 27.4, 29.2, 26.7, 30.9]
}

df = pd.DataFrame(data)
print(df)

# Изпълнение на ANOVA
model = ols('Годност ~ C(Вид) + C(Склад)', data=df).fit() # Важно: добавихме C(), за да сме сигурни, че ги третира като категории
anova_table = sm.stats.anova_lm(model, typ=2)

print("\nANOVA таблица:")
print(anova_table)

   Вид Склад  Годност
0    A     I     33.1
1    A    II     39.0
2    A   III     33.8
3    A    IV     33.0
4    B     I     27.2
5    B    II     36.9
6    B   III     29.7
7    B    IV     28.5
8    C     I     27.4
9    C    II     29.2
10   C   III     26.7
11   C    IV     30.9

ANOVA таблица:
             sum_sq   df         F    PR(>F)
C(Вид)    79.271667  2.0  7.439080  0.023734
C(Склад)  59.936667  3.0  3.749752  0.079070
Residual  31.968333  6.0       NaN       NaN


Открита е зависимост (мултиколинеарност). Използваме само скорост (V).
                            OLS Regression Results                            
Dep. Variable:                      S   R-squared:                       0.966
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     112.7
Date:                Fri, 22 May 2026   Prob (F-statistic):           0.000446
Time:                        08:46:33   Log-Likelihood:                -19.871
No. Observations:                   6   AIC:                             43.74
Df Residuals:                       4   BIC:                             43.33
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------

c:\Users\Anton\anaconda3\Lib\site-packages\statsmodels\regression\linear_model.py:1782: RuntimeWarning: divide by zero encountered in scalar divide
  return 1 - self.ssr/self.centered_tss
c:\Users\Anton\anaconda3\Lib\site-packages\statsmodels\stats\outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)
c:\Users\Anton\anaconda3\Lib\site-packages\statsmodels\stats\stattools.py:74: ValueWarning: omni_normtest is not valid with less than 8 observations; 6 samples were given.
  warn("omni_normtest is not valid with less than 8 observations; %i "


### Разсъждение и обобщение:
Извършен е двуфакторен ANOVA анализ без взаимодействие за определяне влиянието на „Вид консерва“ и „Тип склад“ върху годността на продуктите.

Изводи:
- Вид консерва: Отхвърляме нулевата хипотеза $(p = 0.0237 < 0.05)$, което доказва, че видът на консервата има статистически значим ефект върху срока на годност.
- Тип склад: Нулевата хипотеза не се отхвърля $(p = 0.0791 > 0.05)$, което означава, че типът склад не оказва съществено влияние върху годността при избраното ниво на значимост.
Обобщено, видът на консервата е основният детерминиращ фактор за нейната годност в настоящото изследване.
